text/0000-export.md

```diff
@@ -0,0 +1,401 @@
+
+#### Types with Public Fields
+
+For a simple user defined type where all fields are both public, like the `Point` struct below,
+the relevant parts are the size, alignment, and recursively all field information.
+
+```rust
+#[export]
+#[repr(C)]
+pub struct Point {
+    pub x: f32,
+    pub y: f32,
+    pub name: &str,
+}
+```
+
+The `#[export]` attribute is the user's commitment to keep the type stable, but without `unsafe`,
+any mistakes should _not_ result in unsoundness.
+Accidentally changing the struct to swap the `x` and `name` fields should result in a different hash,
+such that the `f32` won't get interpreted as a `&str`, for example.
+
+Note that, technically, the names of the type and the fields are not relevant, at least _not for memory safety_.
+Swapping the `x` and `y` fields result in surprises and bugs and shouldn't be done,
+but it won't result in undefined behaviour, since any Rust code can swap the fields without using `unsafe`.
+
+However, for public fields, the field names are already part of the stable API, so we include them in the hash as well.
+
+It is an error to use a plain `#[export]` attribute on a type with out stable `#[repr(…)]`,
+if it has any private fields,
+or if any of the fields are not of an `#[export]`ed or builtin type.
+
+#### Types with Private Fields
+
+For types where not all fields are public, the situation is much more complicated.
+
+Private fields usually come with certain *invariants*, and come with `unsafe` code that makes assumptions about them.
+For example, the private fields of a `Vec` are assumed to represent a valid "owned" pointer to an allocation together with its capacity and initialized size.
+
+If it would be possible to define a identically named type with the same fields but different (or no) invariants/assumptions,
+or just change the invariants in an existing library,
+it'd be possible to cause undefined behavior by loading the "wrong" dynamic library.
+
+Therefore, we can't allow a regular `#[export]` attribute on a type with private fields,
+since we have no way of automatically determining the invariants / unsafe assumptions about private fields.
+
+Instead, for these types, we must require the user to *unsafely* commit to
+ABI stability if they want to make the type available to exported functions.
+
+Using `#[export(unsafe_stable_abi = «hash»)]`, one can make the (unsafe) promise
+that the type will remain ABI compatible as long as the provided hash remains the same.
+The hash must have been randomly generated to ensure uniqueness (which is part of the unsafe promise).
+
+```rust
+#[export(unsafe_stable_abi = "ca83050b302bf0644a1417ac3fa6982a")]
+#[repr(C)]
+pub struct ListNode {
+    next: *const ListNode,
+    value: i32,
+}
+```
+
+In this case, using the type as part of a function signature will not result in a hash based on the full (recursive) type definition,
+but will instead be based on the user provided hash (and the size and alignment of the type).
```

Comment:
Is the intention that the regular function signature's hash (commitment) always be derived by fully recursing through the type definition? An alternative would be to structure the commitments as a tree, so the hash (commitment) of an outer type is derived from the hashes (commitments) of its inner types (plus size and alignment). Then public and private types are treated identically. This might mean more hashing overall, but the CPU work done to hash common inner types could be cached and reused across the exported API (which for a large API might actually result in less hashing overall).

Reply:
basically a [Merkle Tree](https://en.wikipedia.org/wiki/Merkle_tree) with a tree node for each type.

Reply:
> Then public and private types are treated identically.

It's critical that they are not treated identically. The safety related invariants of a type with private fields are not fully described by its fields.

E.g. a `struct NonZero { value: i32 }` and `struct NonSeven { value: i32 }` should definitely not be seen as equal.

Similarly, a struct with a private field might be changed over time to have additional safety invariants without changing any of its fields.

And there are also subtle situations where a someone might change the fields of a type without changing its ABI (e.g. changing a `Option<NonNull>` to a raw pointer, changing a `T` to `Wrapping<T>`, moving all fields into an "Inner" struct, etc.).

Reply:
> > Then public and private types are treated identically.
> 
> It's critical that they are not treated identically.

I think you misunderstood my comment; I apologise for not being clearer.

I was not saying that they are treated identically in terms of _how they are hashed_; indeed as you say, the types with private fields have additional considerations, which are represented by (what you call) the hash that the user sets on the type.

I was saying that they are treated identically in terms of _how they are used as children of public parent types_. This would make the logic for calculating the commitment to a public type simpler; there would not be two separate pathways for handling public types (recursing down) and private-containing types (grab the user-pinned hash value). Instead, the logic would always take the hash of the child type, and that hash would either be computed (for public types) or pinned (for private-containing types).